### ISSUE TITLE
Support for other RNGs

### ISSUE THREAD

**Issue by sftrabbit** [bzar/sfxr-rs#12]
Hey there!

I'm in the process of reimplementing [PuzzleScript](https://www.puzzlescript.net/)'s sound effects which use sfxr and was very happy to see that somebody had already implemented sfxr in Rust. However, one small blocker: the methods on `Sample` are hardcoded to use `SmallRng` (and seeds it using `seed_from_u64`), but PuzzleScript uses a different RNG (and seeds it in a different way). It would be great to be able to just pass anything that implements the `Rng` trait.

I'd be happy to submit a PR, but thought I'd ask what your preferred change would be. A couple of ideas:

1. Just break backwards compatibility and make `Sample::explosion`, etc. all take an `Rng`. (And optionally add some new methods that default to using `SmallRng`?)
2. Keep backwards compatibility, and add some new methods like `Sample::explosion_from_rng(rng)`.

I'm relatively new to Rust though, so perhaps there would be a better way to organise this.

**Comment by bzar** [comment]
Hey, nice to see the crate being used! I've had no time to maintain it for some time, but I'll gladly review PRs.

Option 2 is better in my opinion. Breaking API for such an addition is probably not necessary. I'd make the current more complex functions into the `_with_rng` variants and implement the plain ones with that, so `laser` for example just makes and seeds a SmallRng and passes it to `laser_with_rng`.

Since it's not a conversion I would not use `from` in the name to make it more obvious.

**Comment by sftrabbit** [comment]
Thanks for the suggestions! Unfortunately, while trying to make the change, I've realised that I can't implement `Rng` for my PuzzleScript-compatible RNG in such a way that your implementation of sfxr will produce the same sounds as PuzzleScript does.

At least, I think I *could* do it if I changed your `rand_bool` to do `rng.gen::<f32>() * (chance_true + chance_false) > chance_false`, but that would break your backwards compatibility, i.e. a seed from before would now generate different sounds. (There may also be other complications beyond this)

I'm beginning to think that I probably just need a custom fork, due to the specific requirements I have for exactly replicating PuzzleScript's behaviour. Slightly unfortunate, because I was looking forward to making a contribution here, but probably quite sensible, as I shouldn't impose my weird requirements on your crate. Still, nonetheless, thanks for creating this crate so I have something to start with and I'm not doing it all from scratch!
